processing/src/test/java/org/apache/druid/segment/NestedDataColumnIndexerTest.java

```diff
@@ -83,4 +126,459 @@ public void testStuff()
+    List<Cursor> cursorList = cursorSequence.toList();
+    ColumnSelectorFactory columnSelectorFactory = cursorList.get(0).getColumnSelectorFactory();
+
+    ColumnValueSelector valueSelector = columnSelectorFactory.makeColumnValueSelector(NESTED_COL);
+    Assert.assertThrows(
+        UnsupportedOperationException.class,
+        () -> cursorList.get(0).getColumnSelectorFactory().makeDimensionSelector(dimensionSpec)
+    );
+    Assert.assertEquals(StructuredData.wrap("a"), valueSelector.getObject());
+
+    columnSelectorFactory = cursorList.get(1).getColumnSelectorFactory();
+    valueSelector = columnSelectorFactory.makeColumnValueSelector(NESTED_COL);
+    Assert.assertThrows(
+        UnsupportedOperationException.class,
+        () -> cursorList.get(1).getColumnSelectorFactory().makeDimensionSelector(dimensionSpec)
+    );
+    Assert.assertEquals(StructuredData.wrap(2L), valueSelector.getObject());
+    Assert.assertFalse(valueSelector.isNull());
+
+    columnSelectorFactory = cursorList.get(2).getColumnSelectorFactory();
+    valueSelector = columnSelectorFactory.makeColumnValueSelector(NESTED_COL);
+    Assert.assertThrows(
+        UnsupportedOperationException.class,
+        () -> cursorList.get(2).getColumnSelectorFactory().makeDimensionSelector(dimensionSpec)
+    );
+    Assert.assertEquals(StructuredData.wrap(ImmutableMap.of("x", 1.1, "y", 2L)), valueSelector.getObject());
+    Assert.assertFalse(valueSelector.isNull());
+
+    columnSelectorFactory = cursorList.get(3).getColumnSelectorFactory();
+    valueSelector = columnSelectorFactory.makeColumnValueSelector(NESTED_COL);
+    Assert.assertThrows(
+        UnsupportedOperationException.class,
+        () -> cursorList.get(3).getColumnSelectorFactory().makeDimensionSelector(dimensionSpec)
+    );
+    Assert.assertNull(valueSelector.getObject());
+
+    columnSelectorFactory = cursorList.get(4).getColumnSelectorFactory();
+    valueSelector = columnSelectorFactory.makeColumnValueSelector(NESTED_COL);
+    Assert.assertThrows(
+        UnsupportedOperationException.class,
+        () -> cursorList.get(4).getColumnSelectorFactory().makeDimensionSelector(dimensionSpec)
+    );
+    Assert.assertNull(valueSelector.getObject());
+  }
+
+  @Nonnull
+  private static IncrementalIndex makeIncrementalIndex(long minTimestamp)
+  {
+    IncrementalIndex index = new OnheapIncrementalIndex.Builder()
+        .setIndexSchema(
+            new IncrementalIndexSchema(
+                minTimestamp,
+                new TimestampSpec(TIME_COL, "millis", null),
+                Granularities.NONE,
+                VirtualColumns.EMPTY,
+                new DimensionsSpec(Collections.emptyList()),
+                new AggregatorFactory[0],
+                false
+            )
+        )
+        .setMaxRowCount(1000)
+        .setUseNestedColumnIndexerForSchemaDiscovery(true)
+        .build();
+    return index;
+  }
+
+  private MapBasedInputRow makeInputRow(
```

Comment:
Perhaps you could make this `public static` on `MapBasedInputRow`?  Seems like a nice enough helper thing to have available.

Reply:
will move in a future PR